### ISSUE TITLE
EarlyStopping when using an IterableDataset

### ISSUE THREAD

**Issue by ajdillhoff** [PyTorchLightning/pytorch-lightning#814]
I am trying to use an `IterableDataset` while also including early stopping. Using version 0.6.1, it looks like the early stopping callback is only checked after an epoch. When using an `IterableDataset`, I don't see how this is ever called.

I have implemented a quick solution on my local machine after noticing a metric check after the validation loop is run:
https://github.com/PyTorchLightning/pytorch-lightning/blob/4c6c3d04ce7a33a2803971c369b85f1c3e4fd80f/pytorch_lightning/trainer/training_loop.py#L415

My quick fix was to add a call to `EarlyStopping.on_epoch_end` like so:
```Python
# ---------------
# RUN VAL STEP
# ---------------
is_val_check_batch = (batch_idx + 1) % self.val_check_batch == 0
can_check_epoch = (self.current_epoch + 1) % self.check_val_every_n_epoch == 0
should_check_val = (not self.disable_validation and can_check_epoch and
                    (is_val_check_batch or early_stop_epoch))

# fast_dev_run always forces val checking after train batch
if self.fast_dev_run or should_check_val:
    self.run_evaluation(test=self.testing)

    if self.enable_early_stop:
        self.early_stop_callback.check_metrics(self.callback_metrics)
        # CHECK EARLY STOP CALLBACK HERE
        early_stop_epoch = self.early_stop_callback.on_epoch_end(self.current_epoch, self.callback_metrics)
```
Now that `early_stop_epoch` is set after each validation, it is possible to stop using an `IterableDataset`.
```Python
# end epoch early
# stop when the flag is changed or we've gone past the amount
# requested in the batches
if early_stop_epoch or self.fast_dev_run:
    break
```

If there is a native way to trigger the early stopping callback already, then my question is how? If not, then I could submit a PR.

#### What's your environment?

 - OS: Ubuntu 18
 - Packaging: pip
 - Version: 0.6.1

**Comment by williamFalcon** [comment]
i think 0.7.1 fixes this

**Comment by ajdillhoff** [comment]
As of 0.7.5, this issue still persists. Some changes have been made which affect both model checkpointing and early stopping.

# Model Checkpoint Issue
When using an `IterableDataset` that infinitely processes as a single epoch, the model checkpoint callback will never save due. This appears to be intended as seen in `callbacks/model_checkpoint.py:210`

```
if self.epoch_last_check is not None and (epoch - self.epoch_last_check) < self.period:
  # skipping in this term
  return
```

Simply initializing `ModelCheckpoint` with `period=0` allows new checkpoints to be saved as validation is done within a single epoch.

# Early Stopping Issue
Currently, training using an `IterableDataset` means that there is no regular return from `run_training_epoch()` in `trainer/training_loop.py`. Validation checks are still executed starting on line 450:
```
if self.fast_dev_run or should_check_val:
    self.run_evaluation(test_mode=self.testing)
    self.call_checkpoint_callback()
    self.call_early_stop_callback()
```
The early stop callback is called, but nothing is done with its return value. Locally, I have added the following change to fix this issue:
```
early_stop_epoch = self.call_early_stop_callback()
```
If the call to the early stop callback indicates that the training should stop, the training loop breaks and the epoch ends.

It is a small change and I don't see anything else that it could interfere with. However, I have not set up or performed any testing.

I have never submitted a PR before so I am not sure what to do from here.

**Comment by ajdillhoff** [comment]
Any chance that this can be re-opened and addressed?

**Comment by edenlightning** [comment]
@williamFalcon should this be fixed now?
